### PR TITLE
nextcloud:New App - for Nextcloud self hosted environment

### PR DESCRIPTION
**Pull request by iciubotarasu**
Please ensure your pull request (PR) adheres to the following guidelines:

- Please refer to our contributing documentation for any questions on submitting a pull request, link: [Contribution Guide](https://github.com/splunk-soar-connectors/.github/blob/main/.github/CONTRIBUTING.md)

#### Please check if your PR fulfills the following requirements:
- [ ] Testing of all the changes has been performed (for bug fixes / features)
- [x] The readme.html has been reviewed and added / updated if needed (for bug fixes / features)
- [X] Use the following format for the PR description: `<App Name>: <PR Type> - <PR Description>`
- [ ] Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.
- [ ] Verify all checks are passing.
- [ ] Do NOT use the `next` branch of the forked repo. Create separate feature branch for raising the PR.
- [ ] Do NOT submit updates to dependencies unless it fixes an issue.
#### Please check the type of change your PR introduces:
- [x] New App
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation
- [ ] Other (please describe): 

## Security Considerations (REQUIRED)
- [ ] If you are exposing any endpoints using a [REST handler](https://docs.splunk.com/Documentation/SOAR/current/DevelopApps/RESTHandlers), 
  please document them in the `readme.html`.
- [ ] If this is a new connector or you are adding new actions
    - [ ] Please document in the `readme.html` all methods (eg, OAuth) used to authenticate 
      with the service that the connector is integrating with.
    - [ ] If any actions are unable to run on SOAR Cloud, please document this in the `readme.html`.
- [ ] Are you introducing any new cryptography modules? If yes, please elaborate their purpose:
- [ ] Are you are accessing the file system? If yes, please verify that you are only accessing paths returned through 
the [Vault](https://docs.splunk.com/Documentation/SOARonprem/5.2.1/DevelopApps/AppDevAPIRef#Vault) API.
- [ ] Are you are marking code to be ignored by Semgrep with [`nosemgrep`](https://semgrep.dev/docs/ignoring-files-folders-code/#ignoring-code-through-nosemgrep)?
If yes, please provide justification in an additional comment next to the ignored code.

  

## Release Notes (REQUIRED)
- Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.

## What is the current behavior? (OPTIONAL)
- Describe the current behavior that you are modifying.

## What is the new behavior? (OPTIONAL)
- Describe the behavior or changes that are being added by this PR.

## Other information (OPTIONAL)
- Any other information that is important to this PR such as screenshots of how the component looks before and after the change.

## Pay close attention to (OPTIONAL)
- Any specific code change or test case points which must be addressed/reviewed at the time of GA release.

## Screenshots (if relevant)

---
Thanks for contributing!

### CODE REVIEW

**Comment by pdesai-crest** [comment]
Hi @iciubotarasu,
As per our standards, the currently supported "min_phantom_version" should be 6.0.0. Can you please update that and also verify the app on that version?

**Comment by iciubotarasu** [comment]
Hi @pdesai-crest,
I've changed the version and tested it on version 6.0
I also tried to solve most of the FAILED static test results. Not sure how to fix "repo_name_has_expected_app_id".

**Comment by pdesai-crest** [comment]
> Hi @pdesai-crest, I've changed the version and tested it on version 6.0 I also tried to solve most of the FAILED static test results. Not sure how to fix "repo_name_has_expected_app_id".

That static failure is for our internal purpose. We will fix it from our end.

**Comment by pdesai-crest** [comment]
Hi @iciubotarasu , We have made changes as per the our standards. Kindly do not revert it. If you have any concern then please let us know.

**Comment by iciubotarasu** [comment]
@pdesai-crest no worries. I won't change them.

**Comment by pdesai-crest** [comment]
> @pdesai-crest no worries. I won't change them.

Okay, Thank you.

File: requirements.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+requests==2.31.0
```
Comment (rgil-splunk) [inline requirements.txt]: we should remove requests from being explicitly specified in requirements.txt and bundled with apps going forward per https://splunk.slack.com/archives/GBSGHUUD9/p1685364655311529

Comment (pdesai-crest) [reply requirements.txt]: > we should remove requests from being explicitly specified in requirements.txt and bundled with apps going forward per https://splunk.slack.com/archives/GBSGHUUD9/p1685364655311529

Removed the requests library from the requirements.txt
Now can you please approve it?

**Review by rgil-splunk — APPROVED** [review]

File: nextcloud.json [hunk 2]
```diff
@@ -0,0 +1,619 @@
...
+                }
+            ],
+            "render": {
+                "type": "table"
+            },
+            "versions": "EQ(*)"
+        }
+    ],
+    "pip3_dependencies": {
+        "pypi": [
```
Comment (rgil-splunk) [inline nextcloud.json]: we should instead specify this dependency in requirements.txt so it's included as a wheel along with it's subdependencies. that will pin the version and reduce the likelihood of unexpected dependency conflicts in the future

File: nextcloud_connector.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+            file_path = data.get('path')
+        except Exception as e:
+            return action_result.set_status(phantom.APP_ERROR, "There was an issue opening the file: " + str(e))
+        if add_random_string is False:
+            destination_path = param['destination_path'].lstrip('/').rstrip('/') + '/' + name
+        else:
+            destination_path = param['destination_path'].lstrip('/').rstrip('/') + '/' + self._randomstr() + '_' + name
+        result, rest_result = self._make_rest_call(action='upload_file', destination_path=destination_path, file_path=file_path)
+        if result is False:
+            action_result.set_status(phantom.APP_ERROR, "Action failed")
```
Comment (tonyc-phantom) [inline nextcloud_connector.py]: Shouldn't this be a return statement? It looks like it is setting the status to error but will continue on and say it passed.
`return action_result.set_status(phantom.APP_ERROR, "Action failed")`

File: nextcloud_connector.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+        return action_result.set_status(phantom.APP_SUCCESS)
+
+    def _handle_delete_item(self, param):
+        self.save_progress("In action handler for: {0}".format(self.get_action_identifier()))
+        action_result = self.add_action_result(ActionResult(dict(param)))
+        path = param['path'].lstrip('/').rstrip('/')
+        self.save_progress("path: {}".format(path))
+        result, rest_result = self._make_rest_call(action='delete', delete_path=path)
+        if result is False:
+            action_result.set_status(phantom.APP_ERROR, "Action failed")
```
Comment (tonyc-phantom) [inline nextcloud_connector.py]: this should also be a return

File: nextcloud_connector.py [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+    def _handle_download_file(self, param):
+        self.save_progress("In action handler for: {0}".format(self.get_action_identifier()))
+        path = param['path'].lstrip('/').rstrip('/')
+        file_name = path.split('/')[-1]
+        action_result = self.add_action_result(ActionResult(dict(param)))
+        download_path = '/tmp/' + self._randomstr() + file_name
+        self.save_progress("download_path: {}".format(download_path))
+        result, rest_result = self._make_rest_call(action='download_file', download_file=path, download_path=download_path, file_name=file_name)
+        if result is False:
+            action_result.set_status(phantom.APP_ERROR, "Action failed")
```
Comment (tonyc-phantom) [inline nextcloud_connector.py]: Same, should probably check all of the actions.

**Review by tonyc-phantom — CHANGES_REQUESTED** [review]

File: nextcloud_connector.py (hunk 3 above)
Comment (pdesai-crest) [reply nextcloud_connector.py]: done

File: nextcloud_connector.py (hunk 4 above)
Comment (pdesai-crest) [reply nextcloud_connector.py]: done

File: nextcloud_connector.py (hunk 5 above)
Comment (pdesai-crest) [reply nextcloud_connector.py]: done

**Comment by pdesai-crest** [comment]
Hi @iciubotarasu ,
We have made changes  as per the standards. We have remove requests library and its dependencies. In case of failures, added `return` after `if result is False:`

**Review by tonyc-phantom — APPROVED** [review]

**Review by splunk-jessica — APPROVED** [review]

**Review by splunk-jessica — APPROVED** [review]

**Pull request merged**